stablehlo/reference/InterpreterApi.cpp

```diff
@@ -0,0 +1,114 @@
+/* Copyright 2023 The StableHLO Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#include "stablehlo/reference/InterpreterApi.h"
+
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Errc.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/Path.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Support/DebugStringHelper.h"
+#include "stablehlo/reference/Errors.h"
+#include "stablehlo/reference/InterpreterOps.h"
+#include "stablehlo/reference/NumPy.h"
+#include "stablehlo/reference/Ops.h"
+namespace mlir {
+namespace stablehlo {
+namespace {
+func::FuncOp getMainFunction(ModuleOp module, StringRef mainName) {
+  auto functions = module.getOps<func::FuncOp>();
+
+  // If the module has 1 function only, use it as the main function.
+  if (std::distance(functions.begin(), functions.end()) == 1) {
+    return *functions.begin();
+  }
```

Comment:
Nit - there is something odd that this function takes a `mainName` that is ignored if there is only 1 function.

```
func.func @not_main() { ... }
```

If I call `evalModule` for `main` on the above module.. it still runs.

Reply:
Maybe we add a check that the user didn't specify a custom main function. I.e. only use the 1 function if the function we're looking for is `main`.

Let's move this check below the iteration over all ops, and add a check for:

```cpp
  for (auto funcOp : functions)
    if (funcOp.getSymName().equals(mainName)) return funcOp;
 
bool isSingleFunction = std::distance(functions.begin(), functions.end()) == 1;
bool isDefaultLookup = mainName == "main";
if (isSingleFunction && isDefaultLookup)  return *functions.begin();
```